lib/mlton/basic/hash-table.sml

```diff
@@ -1,104 +1,61 @@
-(* Copyright (C) 1999-2006, 2008 Henry Cejtin, Matthew Fluet, Suresh
+(* Copyright (C) 2017 Jason Carr
+ * Copyright (C) 1999-2006, 2008 Henry Cejtin, Matthew Fluet, Suresh
  *    Jagannathan, and Stephen Weeks.
  *
  * MLton is released under a HPND-style license.
  * See the file MLton-LICENSE for details.
  *)
-
-(* This code is not working -- it is not even in sources.cm *)
 structure HashTable: HASH_TABLE =
 struct
 
 structure Set = HashSet
 
-type ('a, 'b) t = ('a * 'b) Set.t
+datatype ('a, 'b) t = T of {set: {hash: word, key: 'a, value: 'b} Set.t,
+                        hash: 'a -> word,
+                        equals: 'a * 'a -> bool,
+                        cache: bool}
```

Comment:
I understand the purpose of `cache`, but I find it an awkward interface.  In particular, the fact that all of the entries have a `cache` field, but it is filled with `0wx0` when `cache = false`.  I see that most of the converted uses of `HashTable` use `cache=true`, so perhaps it would be simpler to always cache.

In ShareZeroVec, I think that `cache=false` arises because `Type.hash` is constant time.  However, I doubt that it is critical to avoid duplicating the hash.